### ISSUE TITLE
Bug: Importing a DialogBlocks project, widgets layout section is not congruent

### ISSUE THREAD

**Issue by rossanoparis** [KeyWorksRW/wxUiEditor#1180]
### Description:

wxUIEditor version: 1.1.3

There are several issues regarding widgets layout that on large projects risk to make the import feature unusefull as it is necessary to check one by one all imported widgets and remake their layout comparing what you see in DialogBlocks ... it is a very hard stuff.

I've made a simple DialogBlocks [project](https://github.com/KeyWorksRW/wxUiEditor/files/12385427/Test-01.zip) to show that.
The issues are related to these properties:
|DB Property|DB XML Project|wxUI Property|wxUI Project|
|-----|---|---|-|
|Border|`<long name="proxy-Border">21</long>`|Border size|always: 5|
|Stretch factor|`<long name="proxy-Stretch factor">1</long>`|proportion|always: 0|
|Align H|`<string name="proxy-AlignH">"Left"</string>`|alignment|always: wxALIGN_CENTER|
|Align V|`<string name="proxy-AlignV">"Top"</string>`|alignment|always: wxALIGN_CENTER|

I think all widgets are affected by this issue during the importing process, perhaps some issues are different depending on the kind of widget, for instance in my example there is a wxStaticLine with the "width" property and horizontal alignment assigned, but during the importing process all is lost regarding its layout.

**Comment by KeyWorksRW** [comment]
Note that supporting proxy-Border conversion to border_size will require #1181 to be completed -- otherwise the layout will look correct on a normal DPI window, but wrong on a high-DPI window.

DialogBlocks will specify `Centre` alignment even if it is invalid under the parent sizer. That effectively overrides `Left` for horizontal alignment. Processing these two values requires looking at the orientation of the parent sizer, and ignoring flags that are invalid for the orientation (e.g., no vertical alignment in a vertically oriented sizer).

**Comment by rossanoparis** [comment]
Thank you for the explanation.

Here we are all very keen regarding the importing features of wxUIEditor, they are indeed the main key able to permit an uge porting of old projects made with DB and FB ... 
Lets say, this issue, it is the last obstacle before starting to think of get rid of the past :)
Do you think this issue could be solved? Or there are some kind of problems able to leed developers to abandon the effort.
We keep fingers crossed.

**Comment by KeyWorksRW** [comment]
The non-scaling fixes went in yesterday as well as fixing the alignment
problems, so they should be in today's (8/21) build. What is *not* in place
yet is adding the additional property and setting that based on what type
of project was imported (DialogBlocks, wxFormBuilder, etc.) as well as for
the user. I'm expecting to have that in place later this week.

**Comment by KeyWorksRW** [comment]
Issues #1181 has been fixed and will show up starting with the 8/23 Daily Build. This should complete the importing problems reported in this issue. I'll leave the issue open for now until the fix is confirmed.

**Comment by rossanoparis** [comment]
thank you @KeyWorksRW I've just used the last build

All reported "problems" in the table above seems solved
It remains an issue regarding the wxStaticLine object
If you try to import this DialogBlocks [project](https://github.com/KeyWorksRW/wxUiEditor/files/12385427/Test-01.zip), the wxStaticLine is aligned to right without "expansion", but the resulting widget in wxUIEditor is correctly align to right, but with the expansion property ON, which wasn't present in DB project

**Comment by KeyWorksRW** [comment]
I just checked in a fix for the static line -- it may affect other controls as well. (the fix will be in the 8/24 Daily Build).

DialogBlocks maintains flags for both horizontal and vertical alignment. It then grays out the ones that cannot be used under the parent sizer's orientation. For example, setting Expand for vertical alignment is ignored if the parent sizer is oriented vertically, but if the parent sizer's orientation is changed, then the horizontal flags are ignored and the vertical flags are used instead.

**wxUiEditor** only stores one set of flags instead of two. Importing now ignores the Expand flag from the DialogBlocks alignment settings that are not used with the current sizer's orientation. That _does _ mean that if the sizer's orientation is changed after importing, then the Expand flag will have to be set in wxUiEditor since the alternate orientation flags are not imported. Personally, I think it makes more sense to set those flags _after_ the sizer orientation is changed since the flag was set in DialogBlocks without knowing what it actually looks like.

**Comment by rossanoparis** [comment]
Thank you @KeyWorksRW, the wxStaticLine widget is fine ...
Below some other issues encountered for the following widgets

**wxRadioBox**
It seems that it is not managed at all.

**wxTextCtrl**
The entire "Styles" (`wxTE_MULTILINE, wxTE_PROCESS_ENTER, ...`) context is not imported.

**wxComboBox**
The entire "Styles" (`wxCB_READONLY, wxCB_SORT, ...`) context is not imported.

**wxStaticText**
|DB Property|DB XML Project|wxUI Property|wxUI Project|
|-----|---|---|-|
|Wrap|`<long name="proxy-Width">250</long>`|Wrap|always: -1|
|Styles|`<bool name="proxy-wxALIGN_CENTRE">1</bool>`|Style|always: wxALIGN_LEFT|

**wxListBox**, **wxCheckListBox**
|DB Property|DB XML Project|wxUI Property|wxUI Project|
|-----|---|---|-|
|Styles|`<bool name="proxy-wxLB_MULTIPLE">1</bool>`|Type|always: wxLB_SINGLE|
|Styles|`<bool name="proxy-wxLB_SORT">1</bool>`|Style|always: none|

**wxListCtrl**
|DB Property|DB XML Project|wxUI Property|wxUI Project|
|-----|---|---|-|
|Styles|`<bool name="proxy-wxLC_ICON">1</bool>`|Mode|always: wxLC_REPORT|
|Styles|`<bool name="proxy-wxLC_AUTOARRANGE">1</bool>`|Style|always: none|

**Comment by Randalphwa** [comment]
Still working on it. I needed to make some underlying changes to make it easier to fix problems from all types of projects that I import, but I've just started flushing out that code, some of which will fix DialogBlocks project issues.

------- Original Message -------

**Comment by KeyWorksRW** [comment]
It looks like all the `SYMBOL_` constants are created internally in DialogBlocks, and not exported to the project file. That means that in DialogBlocks, the main window app is declared as:

```c++
bool Create( wxWindow* parent, wxWindowID id = SYMBOL_CAPPFRAME_IDNAME,
     const wxString& caption = SYMBOL_CAPPFRAME_TITLE,
     const wxPoint& pos = SYMBOL_CAPPFRAME_POSITION,
     const wxSize& size = SYMBOL_CAPPFRAME_SIZE, long style = SYMBOL_CAPPFRAME_STYLE );
```

However, rather than placing the `SYMBOL_` names into the project file, it puts what those symbols are defined as. That means that once imported, the declaration becomes:

```c++
    bool Create(wxWindow *parent, wxWindowID id = ID_CAPPFRAME,
    const wxString& title = _("CAppFrame"),
     const wxPoint& pos = wxDefaultPosition,
     const wxSize& size = wxSize(400, 300), long style = wxDEFAULT_FRAME_STYLE,
     const wxString &name = wxFrameNameStr);
```

In theory, I could probably duplicate those at least for forms like wxFrame, wxDialog, etc. That would make the code generation be closer together, but for general use that doesn't seem a great approach, IMHO. For example, if I'm reading the header file and see `SYMBOL_CAPPFRAME_POSITION` I don't know immediately that it's `wxDefaultPosition` rather then some specfic position. That makes the code a bit less readable. On the flip side, I know you sometimes use more global constants to specify parameter values (the DialogBlocks-generated ones are unique to the form, not global) so it would seem it wouldn't be of value to you. I wanted to check though, in case this was an important importing functionality that you would like to see.

**Comment by KeyWorksRW** [comment]
DialogBlocks has an option called "Fit to content" which determines whether `SetSizeHints()` is called. wxUiEditor doesn't have that option, and typically calls `SetSizerAndFit()` instead. I'm trying to figure out why someone would want the controls at the top of the form hidden and only the bottom ones show up -- until the window is resized, then the top ones start showing up. That seems _really_ unusual behaviour. Normally clipping is done at the bottom and/or right side. Clipping the top controls and only showing them until the _bottom_ of the window is expanded is very strange behaviour.

The test project you attached creates the main window with the top content hidden until the bottom of the window is expanded. My question is whether that's something you would like to see preserved.

**Comment by KeyWorksRW** [comment]
Creating wxPanel as a form is fixed, and layout (horizontal or vertical) in a wxRadioBox is fixed. Also fixed is a bug with spacers not having width/height set, and wxStatusBar not have the number of fields set. I found a few more cases where alignment was generating assertions in wxWidgets which should now be fixed. The fixes will be available in the next daily build.

**Comment by rossanoparis** [comment]
> DialogBlocks has an option called "Fit to content" which determines whether `SetSizeHints()` is called. wxUiEditor doesn't have that option, and typically calls `SetSizerAndFit()` instead. I'm trying to figure out why someone would want the controls at the top of the form hidden and only the bottom ones show up -- until the window is resized, then the top ones start showing up. That seems _really_ unusual behaviour. Normally clipping is done at the bottom and/or right side. Clipping the top controls and only showing them until the _bottom_ of the window is expanded is very strange behaviour.
> 
> The test project you attached creates the main window with the top content hidden until the bottom of the window is expanded. My question is whether that's something you would like to see preserved.

I'm sorry for my late ... 
Regarding what you've written, I think you should not take in consideration DB behaviour as it is a very old RAD tool.
Indeed, you should use an up to date approach instead.

**Comment by rossanoparis** [comment]
Kind @KeyWorksRW today I've used wxUI built yesterday, below some issue still present during importing action.

**widgets**
Here the DB project you can look at: [Test-01.zip](https://github.com/KeyWorksRW/wxUiEditor/files/12605506/Test-01.zip)

wxComboBox:
the wxCB_DROPDOWN flag is not imported

wxListCtrl:
the wxLC_SMALL_ICON flag is not imported (the context "mode" in wxUI remains empty)

wxCheckListBox:
the wxLB_EXTENDED flag is not imported (the context "style" in wxUI is always wxLB_SINGLE)

wxListBox:
the wxLB_EXTENDED flag is not imported (the context "style" in wxUI is always wxLB_SINGLE)

wxStatusBar:
Fields width property is not properly imported, in wxUI it is always -1
It would be nice to have even fileds text as wxUI property to set

**windows**
Here the DB project you can look at: [Test-02.zip](https://github.com/KeyWorksRW/wxUiEditor/files/12605463/Test-02.zip)

Importing DB projects, often inheritated window from custom classes are not imported using the right window kind.
In this case all windows are imported like wxDialog, even though they must be treated as wxPanel, because of their property "Window kind".
Perhaps, you should look at "Window kind" property instead of "Base class" in order to determine the right kind of window.

**Comment by KeyWorksRW** [comment]
I can't reproduce the problem with wxListCtrl -- when I import your file,
the wxLC_SMALL_ICON is set for the mode.

Unfortunately, "Window kind" isn't always part of the class's XML project
file, so I can't rely on it being there. There are actually three different
ways that DialogBlocks specifies what the base wxWidgets class is, and all
three of them can be different from each other. It looks like "Window
kind", if it is specified at all, overrides the other two, but I need to
verify that so I don't risk breaking importing a bunch of other projects in
my test suite.

I have checked in a PR that fixes the styles and status bar widths that you
reported above. That should be in the 9/15 build, assuming I fixed the
problem that broke last night's daily build (😬).

**Comment by KeyWorksRW** [comment]
Actually, this should now be fixed. I was already special-casing wxDialog to check for a wxPanel override from Base class, so I added an additional override from Window kind in case Base class is a custom class. That correctly parsed your second project attachment. I *think* I got the commit in before the daily build kicks off, so this also should be in the 9/15 daily build, and if not definitely the 9/16 build.

**Comment by rossanoparis** [comment]
Thank you very much @KeyWorksRW, all issues reported in this thread has been solved.
Based on this current version (9/15 daily build), I could plan a porting of more complex projects; in case I will report new considerations in this issue.

Another important topic, which I'd like to discuss with you, is related to custom controls which could be present in DB project windows.
Actually they are not imported at all in wxUIEditor.
I'm going to prepare a small example about this topic, after that I will open a specific issue about it.

**Comment by KeyWorksRW** [comment]
Closing this issue as fixed -- other problems with DialogBlocks imports are covered by other issues.
